quinn-proto/src/connection/streams.rs

```diff
@@ -1124,6 +1166,35 @@ impl Recv {
         }
     }
 
+    fn read_chunk(&mut self) -> Result<Option<Bytes>, ReadError> {
+        match self.assembler.read_chunk()? {
```

Comment:
You can implement this via `read_chunks:`

```rust
let mut bufs = &mut [Bytes::new()];
match self.read_chunks(bufs) {
    Ok(Some(1)) => Ok(Some(std::mem::replace(&mut bufs[0], Bytes::new()))),
    Ok(Some(_)) => unreachable!(),
    Ok(None) => Ok(None),
    Err(e) => Err(e),
}
```

Doesn't however really a buy a lot here. Might be more interesting if assembler would be doing less.

Reply:
To me there is no real advantage in factorizing in this way. Since it is different layers composition already, and doing so would add a tiny overhead without making the overall code shorter at all.

Reply:
I think I agree that it doesn't seem to help here.